### PR TITLE
ENH: add rarefy with replacement

### PR DESCRIPTION
**Pull request by stevendbrown**
Adds an option to perform rarefaction with replacement by multinomial sampling.

### CODE REVIEW

**Comment by wasade** [comment]
Attempting to get BIOM 2.1.7 into conda-forge (see conda-forge/biom-format-feedstock#2)

**Comment by stevendbrown** [comment]
The rabbit hole is much deeper than I imagined. Thanks for your efforts @thermokarst, @wasade.

File: q2_feature_table/tests/test_normalize.py [hunk 1]
```diff
@@ -27,6 +27,13 @@ def test_rarefy(self):
         self.assertEqual(set(a.ids(axis='observation')), set(['O1', 'O2']))
         npt.assert_array_equal(a.sum(axis='sample'), np.array([2., 2.]))
 
+    def test_rarefy_replacement(self):
+        t = Table(np.array([[0, 1, 3], [1, 1, 2]]),
+                  ['O1', 'O2'],
+                  ['S1', 'S2', 'S3'])
+        rt = rarefy(t, 3, with_replacement=True)
+        self.assertEqual(rt.shape, (2, 3))
```
[the remark this answers is not in the document]
Comment (stevendbrown) [reply q2_feature_table/tests/test_normalize.py]: Yes; added a loop for a range of subsampling levels.

I agree with your assessment in the lack of clarity around program behavior for samples of insufficient depth. The documentation (`plugin_setup.py`) does describe that samples not meeting the specified depth will be dropped; however, that is unclear in the context of the `with_replacement` switch. So maybe further clarification in the documentation is warranted?

[the remark this answers is not in the document]
Comment (stevendbrown) [reply q2_feature_table/tests/test_normalize.py]: Added text to `plugin_setup.py` clarifying that samples will be dropped for insufficient depth unless subsampling is performed with replacement.

File: q2_feature_table/_normalize.py [hunk 2]
```diff
@@ -9,8 +9,10 @@
 import biom
 
 
-def rarefy(table: biom.Table, sampling_depth: int) -> biom.Table:
-    table = table.subsample(sampling_depth, axis='sample', by_id=False)
+def rarefy(table: biom.Table, sampling_depth: int,
+           with_replacement: bool=False) -> biom.Table:
```
Comment (thermokarst) [inline q2_feature_table/_normalize.py]:
```suggestion
           with_replacement: bool = False) -> biom.Table:
```

Comment (thermokarst) [reply q2_feature_table/_normalize.py]: This is a silly linting error related to the newest version of `flake8`, which is why travis is currently failing, @stevendbrown.

Comment (thermokarst) [reply q2_feature_table/_normalize.py]: To clarify, the error itself is silly, not what you wrote... this new version of `flake8` has generated a lot of busywork for us the last few days!

**Comment by sabahzero** [comment]
I'm unsure if this may be too bold of a comment, but may rarefaction now be obsolete? I've been reading a good deal about the artifacts that can arise with rarefying these past few months and perhaps this is a function worth getting rid of as part of the surmounting statistical evidence?

ie https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531

**Comment by thermokarst** [comment]
Hey there @sabahzero! That is a great question! I suggest moving that discussion over to the QIIME 2 Forum (https://forum.qiime2.org), that is probably a more convenient place to chat about this vs Github comment threads.

**Comment by sabahzero** [comment]
Thank you! I received the notification about the rarefy function and thought it may be relevant within the topic. I went forward with the forum, it currently states my login is pending authentication.

**Comment by thermokarst** [comment]
Hmm, I don't think that is the right forum, since we don't require verification of new user accounts --- did you sign up for the old QIIME 1 forum on accident? The QIIME 2 Forum can be found at https://forum.qiime2.org.

**Comment by sabahzero** [comment]
Used the same link provided before for qiime2, but tried again the approach now and it seemed to work fine. Maybe a glitch? Thanks Matthew!

**Pull request merged**